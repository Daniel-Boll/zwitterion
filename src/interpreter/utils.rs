use std::io::Write;

use super::{AnonymousFunction, Value};

impl std::fmt::Debug for AnonymousFunction {
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    write!(f, "<#closure>")
  }
}

impl std::fmt::Display for Value {
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    match self {
      Value::Int(int) => write!(f, "{}", int),
      Value::Str(string) => write!(f, "{}", string),
      Value::Bool(bool) => write!(f, "{}", bool),
      Value::Tuple(first, second) => write!(f, "({}, {})", first, second),
      Value::Closure(_) => write!(f, "<#closure>"),
    }
  }
}

pub(crate) fn panic<W: Write>(message: String, writer: &mut W) -> ! {
  writer
    .write_all(format!("{}\n", message).as_bytes())
    .unwrap();
  panic!();
}
