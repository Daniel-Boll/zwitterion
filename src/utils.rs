use rinha::rinha;
use std::panic::{self, AssertUnwindSafe};

use crate::interpreter;

pub fn interpret_to_buffer(code: &str) -> String {
  let mut buffer = Vec::new();

  let _ = panic::catch_unwind(AssertUnwindSafe(|| {
    let mut errors = vec![];
    let ast = rinha::FileParser::new()
      .parse(&mut errors, "inline-code", code)
      .unwrap();

    interpreter::from_ast(ast.expression, &mut buffer);
  }));

  String::from_utf8(buffer).unwrap()
}
