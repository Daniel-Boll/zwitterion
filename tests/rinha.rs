use zwitterion::utils::interpret_to_buffer;

#[test]
fn print_number() {
  insta::assert_yaml_snapshot!(interpret_to_buffer(r#"print(1)"#));
}

#[test]
fn print_string() {
  insta::assert_yaml_snapshot!(interpret_to_buffer(r#"print("Hello, world")"#));
}

#[test]
fn sum() {
  insta::assert_yaml_snapshot!(interpret_to_buffer(
    r#"
  let sum = fn(a, b) => { a + b };
  print(sum(1, 2))
  "#
  ));

  insta::assert_yaml_snapshot!(interpret_to_buffer(
    r#"
  print((fn(a, b) => { a + b })(1, 2))
  "#
  ));
}

#[test]
fn tuple_second_fail() {
  insta::assert_yaml_snapshot!(interpret_to_buffer(r#"print(second(true))"#));
}
