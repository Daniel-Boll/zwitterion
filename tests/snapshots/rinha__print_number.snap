---
source: tests/rinha.rs
expression: "interpret_from(r#\"print(1)\"#)"
---
"1\n"

